src/coreclr/src/gc/gc.cpp

```diff
@@ -8276,9 +8347,142 @@ void gc_heap::sort_mark_list()
 
 //    unsigned long start = GetCycleCount32();
 
-    dprintf (3, ("Sorting mark lists"));
-    if (mark_list_index > mark_list)
-        _sort (mark_list, mark_list_index - 1, 0);
+    // compute total mark list size and total ephemeral size
+    size_t total_mark_list_size = 0;
+    size_t total_ephemeral_size = 0;
+    uint8_t* low = (uint8_t*)~0;
+    uint8_t* high = 0;
+    for (int i = 0; i < n_heaps; i++)
+    {
+        gc_heap* hp = g_heaps[i];
+        size_t ephemeral_size = heap_segment_allocated (hp->ephemeral_heap_segment) - hp->gc_low;
+        total_ephemeral_size += ephemeral_size;
+        total_mark_list_size += (hp->mark_list_index - hp->mark_list);
+        low = min (low, hp->gc_low);
+        high = max (high, heap_segment_allocated (hp->ephemeral_heap_segment));
+    }
+
+    // give up if this is not an ephemeral GC or the mark list size is unreasonably large
+    if (settings.condemned_generation > 1 || total_mark_list_size > total_ephemeral_size/256)
```

Comment:
Ok, moved the test for eph GC up and changed it to compare against max_generation.

256 is a number based on a rough back-of-the envelope computation. The idea was that walking the heap will be dominated by fetching cache lines, and for each marked object we will roughly fetch 2 cache lines (one for the marked object, and one for the unmarked object after it). Cache line size is 64 bytes, so if number of mark list elements is ephemeral_size/256, we will fetch about half the cache lines. So at this threshold, there should still be a bit of benefit for using the mark list, but probably not a whole lot, because just walking the heap would benefit from prefetching etc.

So that was my reasoning...